File: GA/GA_Facility_Location_240326/Individual.java

public class Individual {
	int[] bits;    // SOLUTION (OR REPRESENTATION OF A SOLUTION)
	int   fitness; // SOLUTION-QUALITY 
	double p_MUT;  //MUTATION-PROBABILITY of EACH BIT
	int    problemsize = 40;
	
	public Individual(int problemsize){
		//bits    = new int[Daten.anzahlProjekte];
		this.problemsize = problemsize;
		bits    = new int[problemsize];
		p_MUT   = 1./bits.length;
	}

	
	public void mutation(){
		for(int i=0;i<bits.length;i++){
			double p = Math.random();//[0;1)
			if(p < p_MUT){
				if(this.bits[i] == 0)this.bits[i] = 1;
				else                 this.bits[i] = 0;
			}
		}
	}
s
	
	public static void crossover(Individual papa, Individual mama, Individual son, Individual doughter){
		int crosspoint = (int)(Math.random()*papa.bits.length);
		
		
		for(int i=0;i<crosspoint;i++){
			son.bits[i]      = papa.bits[i];
			doughter.bits[i] = mama.bits[i];
		}
		for(int i=crosspoint;i<papa.bits.length;i++){
			son.bits[i]      = mama.bits[i];
			doughter.bits[i] = papa.bits[i];
		}
	}
	
	public void output(){
		for(int i=0;i<bits.length;i++){
			System.out.print(bits[i]);
		}
		System.out.print(" " + fitness);
		System.out.println();
	}

	
	
	public void initialize(){
		int count = 0;
		for(int i=0;i<this.bits.length;i++){
			bits[i] = 0;
			if(Math.random()<0.5){
				bits[i] = 1;
				count++;
			}
		}
		if(count == 0) {
			int nr = (int)(bits.length*Math.random());
			bits[nr] = 1;
		}
	}
	
	public void fitness(){
		fitness = Problem.fitness(bits);
	}

	public void reproduce(Individual template){
		for(int i=0;i<bits.length;i++){
			this.bits[i] = template.bits[i];
		}
		this.fitness = template.fitness;
		this.p_MUT   = template.p_MUT;
	}

	
	
}

File: GA/GA_Sinus_310326/individual.py
from __future__ import annotations

import random
from problem import Problem

class Individual:
    def __init__(self):
        #self.problem_size = problem_size
        self.bits = [0] * Problem.bitstring_length
        self.fitness_value = 0
        self.p_mut = 1.0 / Problem.bitstring_length

    def mutation(self) -> None:
        for i in range(len(self.bits)):
            if random.random() < self.p_mut:
                self.bits[i] = 1 if self.bits[i] == 0 else 0

    @staticmethod
    def crossover(papa: "Individual", mama: "Individual", son: "Individual", daughter: "Individual") -> None:
        crosspoint = random.randrange(0, len(papa.bits))

        for i in range(crosspoint):
            son.bits[i] = papa.bits[i]
            daughter.bits[i] = mama.bits[i]

        for i in range(crosspoint, len(papa.bits)):
            son.bits[i] = mama.bits[i]
            daughter.bits[i] = papa.bits[i]

    def output(self) -> None:
        print("".join(str(bit) for bit in self.bits), self.fitness_value)

    def initialize(self) -> None:
        count = 0
        for i in range(len(self.bits)):
            self.bits[i] = 0
            if random.random() < 0.5:
                self.bits[i] = 1
                count += 1

        #if count == 0:
        #    nr = random.randrange(0, len(self.bits))
        #    self.bits[nr] = 1

    def fitness(self) -> None:
        self.fitness_value = Problem.fitness(self.bits)

    def reproduce(self, template: "Individual") -> None:
        for i in range(len(self.bits)):
            self.bits[i] = template.bits[i]
        self.fitness_value = template.fitness_value
        self.p_mut = template.p_mut
